go/vt/vttablet/tabletserver/repltracker/writer_test.go

```diff
@@ -47,16 +48,222 @@ func TestWriteHeartbeat(t *testing.T) {
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	t.Run("open, heartbeats", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer cancel()
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				assert.EqualValues(t, 0, writeErrors.Get())
+				currentWrites := writes.Get()
+				assert.Greater(t, currentWrites, lastWrites)
+				lastWrites = currentWrites
+			}
+		}
+	})
+}
+
+// TestWriteHeartbeatDisabled tests that the heartbeat writer doesn't write heartbeats when the writer is disabled,
+// but that it does respond to RequestHeartbeats(), and generates heartbeats on a lease.
+func TestWriteHeartbeatDisabled(t *testing.T) {
+	defaultOnDemandDuration = 3 * time.Second
+
+	db := fakesqldb.New(t)
+	defer db.Close()
+
+	tw := newTestWriter(db, nil, tabletenv.Disable, 0)
+
+	// Even though disabled, the writer will have an on-demand duration set.
+	assert.Equal(t, defaultOnDemandDuration, tw.onDemandDuration)
+
+	db.AddQueryPattern("^INSERT INTO.*", &sqltypes.Result{})
+
+	writes.Reset()
+	writeErrors.Reset()
+
+	tw.writeHeartbeat()
+	lastWrites := writes.Get()
+	assert.EqualValues(t, 1, lastWrites)
+	assert.EqualValues(t, 0, writeErrors.Get())
+
+	t.Run("closed, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.Nil(t, rateLimiter)
+	}
+	tw.Open()
+	defer tw.Close()
+	{
+		rateLimiter := tw.safeOnDemandRequestsRateLimiter()
+		assert.NotNil(t, rateLimiter)
+	}
+	t.Run("open, no heartbeats", func(t *testing.T) {
+		<-time.After(3 * time.Second)
+		assert.EqualValues(t, 1, writes.Get())
+	})
+	t.Run("request heartbeats, heartbeats", func(t *testing.T) {
+		tw.RequestHeartbeats()
+		ctx, cancel := context.WithTimeout(context.Background(), tw.onDemandDuration-time.Second)
+		defer cancel()
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				assert.EqualValues(t, 0, writeErrors.Get())
+				currentWrites := writes.Get()
+				assert.Greater(t, currentWrites, lastWrites)
+				lastWrites = currentWrites
+			}
+		}
+	})
+	t.Run("exhaust lease, no heartbeats", func(t *testing.T) {
+		<-time.After(tw.onDemandDuration)
+		currentWrites := writes.Get()
+		<-time.After(3 * time.Second)
```

Comment:
IMO we should use the `onDemandDuration` value again here and everywhere else that we're using the 3 second literal in these tests.

Reply:
It's actually explicitly using a different value. It would be confusing to reuse `onDemandDuration`. What we want to say is: after `onDemandDuration` is expired, no new heartbeats are written, regardless of how many seconds we wait.

If we reused `onDemandDuration`, that could imply to the reader that there is come mysterious connection, where after the duration expires, you need it to expire again so as to eliminate heartbeats.

Dow that make sense?